cmd/slinky/config.go

```diff
@@ -0,0 +1,274 @@
+	// requests.
+	MaxPriceAge time.Duration `json:"maxPriceAge"`
+
+	// Providers is the map of provider names to providers that the oracle will fetch prices from.
+	Providers map[string]config.ProviderConfig `json:"providers"`
+
+	// Production specifies whether the oracle is running in production mode. This is used to
+	// determine whether the oracle should be run in debug mode or not.
+	//
+	// Deprecated: This field is no longer used.
+	Production bool `json:"production"`
+
+	// Metrics is the metrics configurations for the oracle.
+	Metrics config.MetricsConfig `json:"metrics"`
+
+	// Host is the host that the oracle will listen on.
+	Host string `json:"host"`
+
+	// Port is the port that the oracle will listen on.
+	Port string `json:"port"`
+}
+
+func (c *OracleConfig) ValidateBasic() error {
+	if c.UpdateInterval <= 0 {
+		return fmt.Errorf("oracle update interval must be greater than 0")
+	}
+
+	if c.MaxPriceAge <= 0 {
+		return fmt.Errorf("oracle max price age must be greater than 0")
+	}
+
+	for _, p := range c.Providers {
+		if err := p.ValidateBasic(); err != nil {
+			return fmt.Errorf("provider %s is not formatted correctly: %w", p.Name, err)
+		}
+	}
+
+	if len(c.Host) == 0 {
+		return fmt.Errorf("oracle host cannot be empty")
+	}
+
+	if len(c.Port) == 0 {
+		return fmt.Errorf("oracle port cannot be empty")
+	}
+
+	return c.Metrics.ValidateBasic()
+}
+
+func (c *OracleConfig) ToLegacy() config.OracleConfig {
+	providers := make([]config.ProviderConfig, len(c.Providers))
+	var i int
+	for _, providerConfig := range c.Providers {
+		providers[i] = providerConfig
+		i++
+	}
+	return config.OracleConfig{
+		UpdateInterval: c.UpdateInterval,
+		MaxPriceAge:    c.MaxPriceAge,
+		Providers:      providers,
+		Production:     c.Production,
+		Metrics:        c.Metrics,
+		Host:           c.Host,
+		Port:           c.Port,
+	}
+}
+
+func SetDefaults() {
+	setViperDefaultsForDataStructure("", DefaultOracleConfig())
+}
+
+func setViperDefaultsForDataStructure(keyPrefix string, config interface{}) {
+	val := reflect.ValueOf(config)
+	typ := val.Type()
+
+	for i := 0; i < val.NumField(); i++ {
+		field := val.Field(i)
+		jsonFieldTag := typ.Field(i).Tag.Get("json")
+
+		// the fully-qualified key for this field
+		fullKey := keyPrefix + jsonFieldTag
+
+		switch field.Kind() {
+		case reflect.Struct:
+			// set viper defaults for struct via recursion
+			setViperDefaultsForDataStructure(fullKey+jsonFieldDelimiter, field.Interface())
+		case reflect.Map:
+			// set viper defaults for map
+			for _, key := range field.MapKeys() {
+				setViperDefaultsForDataStructure(
+					fullKey+jsonFieldDelimiter+key.String()+jsonFieldDelimiter,
+					field.MapIndex(key).Interface(),
+				)
+			}
+		default:
+			viper.SetDefault(fullKey, field.Interface())
+		}
+	}
+
+	// set the environment prefix
+	viper.SetEnvPrefix("SLINKY_CONFIG")
+	viper.AutomaticEnv()
+	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
```

Comment:
Curious why you're using `_` instead of `.` for env variables.

Reply:
. doesn't work as an env variable for some reason?